### PR TITLE
Topic#29/add blueprint support

### PR DESCRIPTION
**Pull request by FranzPoize**

### CODE REVIEW

**Comment by FranzPoize** [comment]
Can you review that @Adnn ?

File: src/apps/entity_tests/Query_tests.cpp [hunk 1]
```diff
@@ -543,3 +544,60 @@ SCENARIO("Queries on multiple components.")
         }
     }
 }
+
+SCENARIO("Queries on archetype where some entity are blueprints.")
+{
+    GIVEN("An entity manager with 2 entities.")
+    {
+        EntityManager world;
+        std::array<Handle<Entity>, 3> entities{
```
Comment (Adnn) [inline src/apps/entity_tests/Query_tests.cpp]: 2 entities, not 3 (nitpicking)

File: src/apps/entity_tests/Query_tests.cpp [hunk 2]
```diff
@@ -543,3 +544,60 @@ SCENARIO("Queries on multiple components.")
...
+            Query<A>        qA{world};
+
+            THEN("Initially the queries match no entites.")
+            {
+                REQUIRE(qA.countMatches() == 0);
+            }
+
+            WHEN("A component (A) is added on each entity.")
+            {
+                const double a[] = {
```
Comment (Adnn) [inline src/apps/entity_tests/Query_tests.cpp:571]: Also 2

File: src/libs/entity/entity/Archetype.h [hunk 3]
```diff
@@ -389,6 +394,46 @@ Storage<T_component> & Archetype::getStorage(StorageIndex<T_component> aComponen
     return mStores[aComponentIndex]->template as<T_component>();
 }
 
+template<Archetype::Operation N_operation>
```
Comment (Adnn) [inline src/libs/entity/entity/Archetype.h]: Is it used outside of `Archetype.cpp`. If not, it should probably be moved there.

File: src/libs/entity/entity/Entity.cpp [hunk 4]
```diff
@@ -30,6 +30,10 @@ const char * Handle<Entity>::name() const
     return mManager->name(mKey);
 }
 
+const TypeSet Handle<Entity>::getComponentsInfo() const
```
Comment (Adnn) [inline src/libs/entity/entity/Entity.cpp]: I would have reused the existing name `getTypeSet`, for consistency.

File: src/libs/entity/entity/EntityManager.cpp [hunk 5]
```diff
@@ -1,13 +1,50 @@
 #include "EntityManager.h"
 
 #include "Component.h"
+#include "Archetype.h"
+#include "Blueprint.h"
 
+#include <cassert>
```
Comment (Adnn) [inline src/libs/entity/entity/EntityManager.cpp]: `c` last ; )

File: src/libs/entity/entity/Archetype.h [hunk 6]
```diff
@@ -165,6 +170,8 @@ class Archetype
     template <class T_component>
     std::unique_ptr<Archetype> makeRestricted() const;
 
+    std::unique_ptr<Archetype> makeRestrictedFromTypeId(const ComponentId aId) const;
```
Comment (Adnn) [inline src/libs/entity/entity/Archetype.h]: Is the new dynamic overload used? (or is it for client consumption?)

File: src/libs/entity/entity/detail/QueryBackend.h [hunk 7]
```diff
@@ -208,7 +210,7 @@ void QueryBackend<VT_components...>::pushIfMatches(const TypeSet & aCandidateTyp
                                                    const ArchetypeStore & aStore)
 {
     if(std::includes(aCandidateTypeSet.begin(), aCandidateTypeSet.end(),
-                     GetTypeSet().begin(), GetTypeSet().end()))
+                     GetTypeSet().begin(), GetTypeSet().end()) && std::find(aCandidateTypeSet.begin(), aCandidateTypeSet.end(), getId<Blueprint>()) == aCandidateTypeSet.end())
```
Comment (Adnn) [inline src/libs/entity/entity/detail/QueryBackend.h]: `\n`

File: src/libs/entity/entity/EntityManager.cpp [hunk 8]
```diff
@@ -1,13 +1,50 @@
...
 #include <iterator>
 
 
 namespace ad {
 namespace ent {
 
+Handle<Entity> EntityManager::createFromBlueprint(Handle<Entity> aBlueprint, const char * aName)
```
Comment (Adnn) [inline src/libs/entity/entity/EntityManager.cpp:14]: Maybe we can extract a sort of `Entity::copy()` for the API, and use it to implement as a copy then remove.

**Review by Adnn — CHANGES_REQUESTED** [review]

**Pull request merged**